alacritty/src/event.rs

```diff
@@ -783,12 +782,15 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
     }
 
     /// Paste a text into the terminal.
-    fn paste(&mut self, text: &str) {
+    fn paste(&mut self, text: &str, bracketed: bool) {
         if self.search_active() {
             for c in text.chars() {
                 self.search_input(c);
             }
-        } else if self.terminal().mode().contains(TermMode::BRACKETED_PASTE) {
+        } else if bracketed && self.terminal().mode().contains(TermMode::BRACKETED_PASTE) {
+            self.on_typing_start();
+            self.clear_selection();
```

Comment:
What's the purpose to ignore the early return to have less duplicate code? The check for non-zero before scrolling was pulled to not damage entire screen, but I there could be a logic inside the scrolling itself to not damage when nothing changed, I don't remember.

Reply:
There is no early return here. What are you talking about?

Reply:
I'm talking about, why you don't use early return to remove like 4 lines of duplicates.

Reply:
Early return where exactly?

Reply:
I've posted the original diff with the early return though, and all the code needed.

https://github.com/alacritty/alacritty/pull/6849#discussion_r1161350483

The only issue is that `NOTE` could be extracted, so it's the same in `received_char` and `paste`.

Reply:
That original code is much worse than what I've implemented here though.

Reply:
I just don't like that we have two else branches with near the same code right now, which you could call right before taking both of the branches, this is just error prone when you update only one and forget about the other. I'd maybe even refactor this function logic out and use the same in `received_char` as well, to not break it.

Reply:
Your suggestion just makes things more verbose and more fragile by increasing the control-flow complexity though.

Reply:
I just had it before the branching before, which behaved the same way as far as I could tell. But you wanted to change that so this is imo the best alternative.

Reply:
Maybe something like this then:

```diff
diff --git a/alacritty/src/event.rs b/alacritty/src/event.rs
index 70c88a52..6c34ea61 100644
--- a/alacritty/src/event.rs
+++ b/alacritty/src/event.rs
@@ -692,6 +692,17 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
         }
     }
 
+    /// Handle keyboard typing into the terminal.
+    ///
+    /// This function must be called when the user typing results in writing to the
+    /// pty.
+    #[inline]
+    fn on_terminal_typing_start(&mut self) {
+        self.on_typing_start();
+        self.scroll(Scroll::Bottom);
+        self.clear_selection();
+    }
+
     /// Process a new character for keyboard hints.
     fn hint_input(&mut self, c: char) {
         if let Some(hint) = self.display.hint_state.keyboard_input(self.terminal, c) {
@@ -788,8 +799,7 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
                 self.search_input(c);
             }
         } else if bracketed && self.terminal().mode().contains(TermMode::BRACKETED_PASTE) {
-            self.on_typing_start();
-            self.clear_selection();
+            self.on_terminal_typing_start();
 
             self.write_to_pty(&b"\x1b[200~"[..]);
 
@@ -802,11 +812,8 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
             self.write_to_pty(filtered.into_bytes());
 
             self.write_to_pty(&b"\x1b[201~"[..]);
-
-            self.scroll(Scroll::Bottom);
         } else {
-            self.on_typing_start();
-            self.clear_selection();
+            self.on_terminal_typing_start();
 
             // In non-bracketed (ie: normal) mode, terminal applications cannot distinguish
             // pasted data from keystrokes.
@@ -815,8 +822,6 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
             // task to solve in a general way), we'll just replace line breaks (windows and unix
             // style) with a single carriage return (\r, which is what the Enter key produces).
             self.write_to_pty(text.replace("\r\n", "\r").replace('\n', "\r").into_bytes());
-
-            self.scroll(Scroll::Bottom);
         }
     }
 
diff --git a/alacritty/src/input.rs b/alacritty/src/input.rs
index 4e256672..95a9e207 100644
--- a/alacritty/src/input.rs
+++ b/alacritty/src/input.rs
@@ -116,6 +116,7 @@ pub trait ActionContext<T: EventListener> {
     fn search_direction(&self) -> Direction;
     fn search_active(&self) -> bool;
     fn on_typing_start(&mut self) {}
+    fn on_terminal_typing_start(&mut self) {}
     fn toggle_vi_mode(&mut self) {}
     fn hint_input(&mut self, _character: char) {}
     fn trigger_hint(&mut self, _hint: &HintMatch) {}
@@ -993,12 +994,7 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
             return;
         }
 
-        self.ctx.on_typing_start();
-
-        if self.ctx.terminal().grid().display_offset() != 0 {
-            self.ctx.scroll(Scroll::Bottom);
-        }
-        self.ctx.clear_selection();
+        self.ctx.on_terminal_typing_start();
 
         let utf8_len = c.len_utf8();
         let mut bytes = vec![0; utf8_len];

```

I'm just worried, because we already had in the past breakages in terminal input due to inconsistent handling across the code base, such as, where latency regressed by 1 frame due to dirty bit altering early on or not doing various checks.